### PR TITLE
Simplify Arg_spec

### PR DESCRIPTION
**Pull request by snowleopard**
This is an attempt to simplify `Arg_spec` as a preparation for replacing arrows with selective functors.

The main change is the constructor `Arg_spec.Dyn`, which now takes `static t Build.s` instead of a function `'a -> (unit, static) t`, which allows us to get rid of the first type parameter `'a` in `('a, 'b) Arg_spec.t`, and bring us closer to replacing arrows with applicative and selective functors.

As part of this work we plan to rename `Arg_spec` to `Command`, because we need to move `run` into this module, and `Command.run` makes more sense than `Arg_spec.run`.

I've managed to convert a few modules to `Command` from `Arg_spec`, but there is still a lot of work. This PR is not ready yet for review, but any comments are welcome.

/cc @diml @aalekseyev

### CODE REVIEW

**Comment by snowleopard** [comment]
OK, `Arg_spec` is now gone, and we are much closer to replacing arrows with applicative and selective functors. Please review.

**Comment by aalekseyev** [comment]
Build starts failing at 06883c7, I'm trying to understand why...

**Comment by snowleopard** [comment]
@aalekseyev Interesting, `dune build` was working fine for me locally for every commit, but now I tried to do the full bootstrapping build, starting with `make`, and the build failed with a similar `bench` error.

File: src/coq_rules.ml [hunk 1]
```diff
@@ -66,31 +65,33 @@ let setup_rule ~expander ~dir ~cc ~source_rule ~coq_flags ~file_flags
...
 
   (* Rules for the files *)
   [coqdep_rule;
-   deps_of >>>
-   Expander.expand_and_eval_set expander coq_flags ~standard:(Build.return []) >>>
-   Build.run ~dir cc.coqc (Dyn (fun flags -> As flags) :: cc_arg)
+   Build.S.seq deps_of (
+    let coq_flags = Expander.expand_and_eval_set expander coq_flags ~standard:(Build.return []) in
+    let coq_flags = Build.S.map coq_flags ~f:(fun l -> Command.As l) in
+    Command.run ~dir cc.coqc (Command.Dyn coq_flags :: cc_arg))
```
Comment (snowleopard) [inline src/coq_rules.ml]: TODO: use `Command.dyn_args coq_flags :: cc_arg` instead, which makes the above `Build.S.map` unnecessary.

File: src/coq_rules.ml [hunk 2]
```diff
@@ -127,11 +128,11 @@ let setup_ml_deps ~lib_db libs =
   let ml_iflags, mlpack =
     let libs = libs_of_coq_deps ~lib_db libs
                |> Lib.closure ~linking:false |> Result.ok_exn in
-    Util.include_flags libs, List.concat_map ~f:ml_pack_files libs
+      Util.include_flags libs, List.concat_map ~f:ml_pack_files libs
```
Comment (snowleopard) [inline src/coq_rules.ml]: TODO: Undo.

File: src/coq_rules.ml [hunk 3]
```diff
@@ -110,7 +111,7 @@ let libs_of_coq_deps ~lib_db libs =
   Result.List.map ~f:(Lib.DB.resolve lib_db) libs |> Result.ok_exn
 
 (* compute include flags and mlpack rules *)
-let setup_ml_deps ~lib_db libs =
+let setup_ml_deps ~lib_db libs : _ Command.t * _ =
```
Comment (snowleopard) [inline src/coq_rules.ml]: TODO: Undo.

File: src/js_of_ocaml_rules.ml [hunk 4]
```diff
@@ -99,31 +98,32 @@ let jsoo_archives ~ctx lib =
...
-      let all_libs = List.concat_map libs ~f:(jsoo_archives ~ctx) in
-      (* Special case for the stdlib because it is not referenced in the META *)
-      let all_libs =
-        in_build_dir ~ctx ["stdlib"; "stdlib.cma.js"] :: all_libs in
-      let all_other_modules =
-        List.map cm ~f:(fun m -> Path.extend_basename m ~suffix:".js")
-      in
-      Arg_spec.Deps (List.concat [all_libs; all_other_modules]))
+  let get_all = Build.S.map cm ~f:(
+    fun cm ->
```
Comment (snowleopard) [inline src/js_of_ocaml_rules.ml]: TODO: Fix alignment to make the diff smaller.

File: src/lib_rules.ml [hunk 5]
```diff
@@ -304,19 +296,16 @@ module Gen (P : sig val sctx : Super_context.t end) = struct
...
-        >>>
-        Build.run ~dir:ctx.build_dir
-          (Ok ocamlopt)
-          [ Dyn (fun flags -> As flags)
-          ; A "-shared"; A "-linkall"
-          ; A "-I"; Path dir
-          ; A "-o"; Target dst
-          ; Dep src
-          ]
+        Build.S.seq (Build.dyn_paths (Build.arr (fun () ->
```
Comment (snowleopard) [inline src/lib_rules.ml]: TODO: Fix alignment to make the diff smaller.

File: src/module_compilation.ml [hunk 6]
```diff
@@ -202,20 +199,21 @@ let ocamlc_i ?sandbox ?(flags=[]) ~dep_graphs cctx (m : Module.t) ~output =
...
-       ; (match CC.alias_module cctx with
-          | None -> S []
-          | Some (m : Module.t) ->
-            As ["-open"; Module.Name.to_string (Module.name m)])
-       ; As flags
-       ; A "-short-paths"
-       ; A "-i"; Ml_kind.flag Impl; Dep src
-       ]
-     >>^ (fun act -> Action.with_stdout_to output act)
+    (Build.S.seq cm_deps
```
Comment (snowleopard) [inline src/module_compilation.ml]: TODO: Fix alignment to make the diff smaller.

File: src/ocamlobjinfo.mli [hunk 7]
```diff
@@ -9,7 +9,7 @@ val rules
   :  dir:Path.t
   -> ctx:Context.t
   -> unit:Path.t
-  -> (_, Action.t) Build.t * (_, t) Build.t
+  -> Action.t Build.s * t Build.s
```
Comment (snowleopard) [inline src/ocamlobjinfo.mli]: TODO: This looks like an unnecessary change at this stage.

File: src/preprocessing.ml [hunk 8]
```diff
@@ -394,21 +394,20 @@ let build_ppx_driver sctx ~dep_kind ~target ~dir_kind ~pps ~pp_names =
...
      Build.write_file_dyn ml);
   add_rule
-    (Build.record_lib_deps
-       (Lib_deps.info ~kind:dep_kind (Lib_deps.of_pps pp_names))
-     >>>
-     Build.of_result_map driver_and_libs ~f:(fun (_, libs) ->
-       Build.paths (Lib.L.archive_files libs ~mode))
-     >>>
-     Build.run (Ok compiler) ~dir:ctx.build_dir
+    (Build.S.seqs
```
Comment (snowleopard) [inline src/preprocessing.ml]: TODO: Try to make the diff smaller.

**Comment by aalekseyev** [comment]
Sorry, I meant to say that tests start failing rather than build. 
If you want to run tests locally you'd need to install a bunch of stuff, something like:

```
    opam install \
         ocamlfind \
         utop \
         ppxlib \
         odoc.1.4.0 \
         menhir \
         ocaml-migrate-parsetree \
         js_of_ocaml-ppx \
         js_of_ocaml-compiler \
         coq
```
(the list taken from .travis-ci.sh)

Only a subset of these are necessary for the tests run by [make test] though (pretty sure coq isn't).

**Comment by snowleopard** [comment]
> Sorry, I meant to say that tests start failing rather than build.

Are you talking about the CI testsuite or about your local tests? To me it seems like a build failure rather than test failure, e.g. see this CI job: https://travis-ci.org/ocaml/dune/jobs/536585685, which fails with

```
+./_boot/install/default/bin/dune build @runtest-no-deps
+cat runtest-no-deps.out
File "bench/dune_bench/dune", line 5, characters 18-27:
5 |  (preprocess (pps ppx_bench)))
                      ^^^^^^^^^
Error: Library "ppx_bench" not found.
Hint: try: dune external-lib-deps --missing @runtest-no-deps
+exit 1
```

> If you want to run tests locally you'd need to install a bunch of stuff, something like:

Great, thanks, will give it a try.

File: src/command.ml [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+      | Paths fns ->
+        List.map fns ~f:(Path.reach ~from:dir)
+      | S ts -> List.concat_map ts ~f:loop_static
+      | Concat (sep, ts) -> [String.concat ~sep (loop_static (S ts))]
+      | Hidden_deps l ->
+        static_deps := Dep.Set.union !static_deps l;
+        []
+      | Fail f -> f.fail ()
+    in
+    (loop_static t, !static_deps)
```
Comment (aalekseyev) [inline src/command.ml]: Argh, it took ages for me to find this. This dereferences `!static_deps` and runs the loop in unspecified order.
You want to write:
```
let res = loop_static t in
(res, !static_deps)
```

Comment (snowleopard) [reply src/command.ml]: Ouch! Thanks for spotting -- fixing in the next commit.

File: src/command.mli [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+  | Fail     : fail -> _ t
+
+(* Create a dynamic command line arguments. *)
+val dyn_args : string list Build.s -> dynamic t
+
+(* Create a dynamic dependencies. *)
+val dyn_dep  : Path.t Build.s -> dynamic t
+val dyn_deps : Path.t list Build.s -> dynamic t
+
+val add_targets : dynamic t list -> Path.t list -> Path.t list
```
Comment (diml) [inline src/command.mli]: I think this can be deleted. It was used to implement `Build.run` when it was leaving in `build.ml`

File: src/command.mli [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+  | Hidden_targets : Path.t list -> dynamic t
+  | Dyn      : static t Build.s -> dynamic t
+  | Fail     : fail -> _ t
+
+(* Create a dynamic command line arguments. *)
+val dyn_args : string list Build.s -> dynamic t
+
+(* Create a dynamic dependencies. *)
+val dyn_dep  : Path.t Build.s -> dynamic t
+val dyn_deps : Path.t list Build.s -> dynamic t
```
Comment (diml) [inline src/command.mli]: To be consistent with similar functions in other modules, these function should take `Dep.t` values rather than `Path.t` ones. Alternatively they could be renamed. Although, given that each of these is used only once in the rest of the code base, I would just inline them.

File: src/command.mli [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+    difference between the two is that [A s] produces exactly the
+    argument [s], while [Path p] produces a string that depends on
+    where the command is executed. For instance [Path (Path.of_string
+    "src/foo.ml")] will translate to "../src/foo.ml" if the command is
+    started from the "test" directory.  *)
+
+type static = Static
+type dynamic = Dynamic
+
+type _ t =
```
Comment (diml) [inline src/command.mli]: We should put this in a sub-module `Args`. `Command.t` makes me think of something that represent a command. But we can do this in a sub-sequence PR.

File: src/command.mli (hunk 11 above)
Comment (aalekseyev) [reply src/command.mli]: These can't take an arbitrary `Dep.t` here because we're adding the path both as a dependency and as an argument on the command line. This confusion seems very hard to avoid in this module in general.

I'm OK with inlining though.

File: src/command.mli (hunk 12 above)
Comment (aalekseyev) [reply src/command.mli]: Indeed. I like `Command.Args.t` better, but you know what name might be even better? `Arg_spec.t` :-)

**Comment by aalekseyev** [comment]
I found and fixed another bug (again it was a side-effect running at a wrong time: 79f824b) and rebased this PR such that all commits pass tests. 

I'm rebasing it onto master now.

@snowleopard, CI complains that you didn't declare that you're releasing your contributions under an appropriate license. The way to do that is to add "Signed-off-by: Blah Blah <blah.blah@gmail.com>" to your commit message.
(I find it strange that the message doesn't point to the license file or anything like that though)

@diml, what should be done here? What's the process for new contributors? Should @snowleopard  just add this to all his commit messages?

**Comment by aalekseyev** [comment]
By the way, we shouldn't try to preserve preserve the old code indentation.

`patdiff` can handle alignment changes pretty well and most of us review with patdiff, not with github.

In fact I reindented most of the files with ocp-indent now and we normally keep them that way.

File: src/command.mli (hunk 11 above)
Comment (snowleopard) [reply src/command.mli]: I'm fine with inlining too. Will do that.

File: src/command.mli (hunk 12 above)
Comment (snowleopard) [reply src/command.mli]: Happy to switch to `Command.Args.t` or back to `Arg_spec.t` :-) Let me know what you prefer.

**Comment by snowleopard** [comment]
@diml @aalekseyev Thanks for the review and fixes! 

> By the way, we shouldn't try to preserve preserve the old code indentation.

Got it! I've resolved the corresponding TODOs.

There are still a few TODOs left:

- [ ] Switch from `Command.t` to `Command.Args.t`.
- [ ] Inline `dyn_dep` and `dyn_deps`.
- [ ] Do not export `add_targets`.
- [ ] Add the "Signed-off-by" bit. Not sure if it's OK to do it only for the final commit? I guess this PR will be squashed anyway.

Happy to do them today.

**Comment by aalekseyev** [comment]
For CI to be happy you need to sign-off all commits, but indeed if you squash it then that becomes easy.

That said, I've used this script to sign off all un-signed-off commits before (it has my name hard-coded):

```
git filter-branch -f --msg-filter 'srcz=$(cat); printf "%s\n" "$srcz"; if ! (printf "%s\n" "$srcz" | grep -q "Signed-off"); then echo; echo "Signed-off-by: Arseniy Alekseyev <aalekseyev@janestreet.com>"; fi' origin/master..HEAD
```

**Comment by aalekseyev** [comment]
@snowleopard, ah, sorry I broke `js_of_ocaml_rules.boot.ml` and thanks for fixing.

File: src/command.mli (hunk 11 above)
Comment (snowleopard) [reply src/command.mli]: Done.

File: src/command.mli (hunk 10 above)
Comment (snowleopard) [reply src/command.mli]: Done.

File: src/command.mli (hunk 12 above)
Comment (snowleopard) [reply src/command.mli]: Done.

**Comment by snowleopard** [comment]
OK, I fixed everything, and all my commits are now signed off thanks to @aalekseyev's script.

**Comment by snowleopard** [comment]
CI fails with:

```diff
--- test/blackbox-tests/test-cases/js_of_ocaml/run.t	2019-05-27 19:49:31.235901158 +0000
+++ test/blackbox-tests/test-cases/js_of_ocaml/run.t.corrected	2019-05-27 19:49:50.908024694 +0000
@@ -7,11 +7,11 @@
         ocamlc lib/.x.objs/byte/x__.{cmi,cmo,cmt}
         ocamlc lib/.x.objs/byte/x__Y.{cmi,cmo,cmt}
       ocamlopt lib/.x.objs/native/x__Y.{cmx,o}
+   js_of_ocaml bin/technologic.bc.runtime.js
            ppx bin/technologic.pp.ml
       ocamldep bin/.technologic.eobjs/technologic.pp.ml.d
            ppx bin/z.pp.ml
       ocamldep bin/.technologic.eobjs/z.pp.ml.d
-   js_of_ocaml bin/technologic.bc.runtime.js
    js_of_ocaml .js/js_of_ocaml/js_of_ocaml.cma.js
            ppx lib/x.pp.ml
       ocamldep lib/.x.objs/x.pp.ml.d
```

Looks like I managed to mess up the order of `js_of_ocaml` arguments.

**Comment by aalekseyev** [comment]
That's the order things are built in, not the order of arguments. I half expected this to change because we're at least changing where the Built.t that computes the dynamic Arg_spec occurs in the arrow representation.

I think we should just accept this change and move on. I pushed the commit that does it.

(the reason I missed it before is that this test is not built in @runtest alias, only in @runtest-js, to build which you additionally need at least node-js installed and `opam install js_of_ocaml-compiler js_of_ocaml-ppx`)

**Comment by snowleopard** [comment]
Thanks @aalekseyev! All green now.

**Comment by aalekseyev** [comment]
I merged this. Thanks @snowleopard.

**Comment by snowleopard** [comment]
@aalekseyev Awesome, thanks for your help! I plan to send another PR soon, completing the switch from arrows to selective functors.

**Comment by diml** [comment]
> @diml, what should be done here? What's the process for new contributors? Should @snowleopard just add this to all his commit messages?

Yh, you can simply ask new contributors to sign their commit as per CONTRIBUTING.md. When they sign, they assert that they read and agreed with the DCO.
